nixos/postgresql: improve local peer authentication with default map

wolfgangwalther opened the pull request:
This allows to easily map allowed database roles to system users.

Background: As mentioned elsewhere, I'm currently adding a NixOS module for pgBackRest. I'll need to give pgBackRest superuser access to the local PostgreSQL cluster. To do so, I have 3 different options:
- Run pgBackRest as the `postgres` system user. This is not satisfying, because it would have write access on the filesystem level, which is discouraged. Read-only should be enough.
- Add a `pgbackrest` database role matching the `pgbackrest` system user. This quickly leads to #346886 during recovery and #403645 to solve it... and we're not sure whether we want to do that.
- Add a map between the `pgbackrest` system and the `postgres` database user. To do that, I would need to modify the `authentication` / `pg_hba.conf` part via pgbackrest's module. This is annoying, because it might lead to conflicts with user defined rules.

It's better to use a map for local peer authentication by default, which can then be easily extended upon by lines in `identMap`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
nixos/postgresql: improve local peer authentication with default map

This allows to easily map allowed database roles to system users.

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -676,12 +683,19 @@ in
       (mkAfter ''
         # default value of services.postgresql.authentication
-        local all all              peer
+        local all all              peer map=nixos
```
What implications would `all` have?

Ma27 reviewed (commented):
Admittedly, I kinda grew up with postgresql on NixOS, so I don't remember having actively touched this feature. Is there a reason we want this over the current state now that we're in agreement on #403645?

Also, what's the benefit of embedding this into the module over letting modules write their own map as Hydra does it?

Regarding

> This is annoying, because it might lead to conflicts with user defined rules.

I think what we do in there should we well-documented so people can spot conflicts like that. Additionally, it may be an option to do some RFC42-like stuff here?

wolfgangwalther committed:
nixos/hydra: avoid conflicts for local postgres ident map

The key change here is, that the match was on the hydra database only
previously, but is now limited to to the hydra role instead. This avoids
conflicts with rules that are created by other modules or downstream
users.

With this change, we can remove the additional "postgres postgres" line,
because the default pg_hba line will kick in again and allow the
postgres user access to the postgres role.

Renaming the map from hydra-users to hydra is for consistency, so that
all modules can define maps matching in name with the role they manage.

The change from ident to peer is just cosmetic, ident is only used for
TCP connections and falls back to peer anyway.

Ma27 commented:
> have each module define their own map, but instead of targeting a specific database like the hydra module does it, those should target a specific database role

Does that mean that _every_ module has to do this? Because, right now local unix auth with is failing for everything except postgres and this definitely has a pretty large fallout.

> WDYT? Is that the direction you're thinking, too?

Ah yeah, that's a good idea. For me, this is mostly a way to control the usage of this in nixpkgs to a certain degree (see the problems of the Hydra module).

If others have special rules or prefer to write the verbatim variant, that's fine IMHO.

Ma27 commented:
> WDYT about doing this more structued way in a second step, after branch-off, too? I'd like to have a bit more room to get the design of those options right, also considering how to proceed with ensureUsers in the future etc.

Ah damn it, I knew I forgot about something in my previosu comment!
Yes, please. No need to rush on this.

> If you agree with that, what kind of documentation do you think we need for the current change in this PR?

Given that some of the findings were more or less discovered while discussing this change, I think the recommendation on how to write hba rules with maps (i.e. restrict by user not by dbname) should absolutely go into the manual.

Perhaps even a short sentence into the release notes?

Ma27 commented on nixos/modules/services/databases/postgresql.md:
```diff
@@ -170,6 +170,41 @@ are already created.
+System users and database roles might not always match.
+In this case, to allow access for a service, you have two options:
+
+  1. Create a database role matching your system user during [initialization](#module-services-postgres-initializing).
```
I think it's a little odd to suggest creating a role matching yoru system user if the precondition is that those don't match ;)

Ma27 reviewed (commented):
One remark about the docs, other than that, LGTM :+1:

Ma27 commented:
> I couldn't come up with anything, really. To me, it doesn't feel like we're really adding anything here, we're mostly documenting a best-practice that doesn't lead to conflicts.

Given that I don't see a case where this is breaking, I agree.
